Add `CloudFront-TLS-1-2-2021-ChaCha20-Boosted` Security Policy w/ Docs Update (#3686)
aws/s2n-tls: docs/USAGE-GUIDE.md

The "rfc9151" security policy is derived from [Commercial National Security Algorithm (CNSA) Suite Profile for TLS and DTLS 1.2 and 1.3](https://datatracker.ietf.org/doc/html/rfc9151).

s2n-tls does not expose an API to control the order of preference for each ciphersuite or protocol version. Unless [ChaCha20-Poly1305 boosting](#chart-security-policy-version-to-protocol-version-and-ciphersuites-with-chacha20-poly1305-boosting-support) is enabled then s2n-tls follows the following order:

*NOTE*: All ChaCha20-Poly1305 cipher suites will not be available if s2n-tls is not built with an Openssl 1.1.1 libcrypto. The underlying encrypt/decrypt functions are not available in older versions.

1. Always prefer the highest protocol version supported
2. Always use forward secrecy where possible. Prefer ECDHE over DHE.
3. Prefer encryption ciphers in the following order: AES128, AES256, ChaCha20, 3DES, RC4.
4. Prefer record authentication modes in the following order: GCM, Poly1305, SHA256, SHA1, MD5.

> ### Chart: Security Policy Version To Protocol Version And Ciphersuites (With ChaCha20-Poly1305 Boosting Support)

Author: How would we organise this when more boosted security policies come? Do we just put them in the original big chart?

The "rfc9151" security policy is derived from [Commercial National Security Algorithm (CNSA) Suite Profile for TLS and DTLS 1.2 and 1.3](https://datatracker.ietf.org/doc/html/rfc9151).

s2n-tls does not expose an API to control the order of preference for each ciphersuite or protocol version. Unless [ChaCha20-Poly1305 boosting](#chacha20-poly1305-boosting) is enabled then s2n-tls follows the following order:

> 

Reviewer: This order isn't for run-time cipher selection logic, it's for the general order cipher suites are listed in security policies. I would not mention chacha20 boosting here.

*NOTE*: All ChaCha20-Poly1305 cipher suites will not be available if s2n-tls is not built with an Openssl 1.1.1 libcrypto. The underlying encrypt/decrypt functions are not available in older versions.

1. Always prefer the highest protocol version supported
2. Always use forward secrecy where possible. Prefer ECDHE over DHE.
3. Prefer encryption ciphers in the following order: AES128, AES256, ChaCha20, 3DES, RC4.
4. Prefer record authentication modes in the following order: GCM, Poly1305, SHA256, SHA1, MD5.

#### ChaCha20-Poly1305 Boosting

> The prioritization of ChaCha20 ciphersuites is known as ChaCha20 boosting. An s2n-tls server will prioritize its ChaCha20 ciphersuites over its non-ChaCha20 ciphersuites if the server configures a security policy with ChaCha20 boosting enabled AND the client sends a ChaCha20 ciphersuite as its most preferred ciphersuite. 

Reviewer: As a customer, this doesn't tell me whether or not I want ChaCha20 boosting. You need to explain what this feature is and why I would want it.

Talk about the use case where some clients, particularly mobile clients, don't have AES optimized and so will prefer Chacha20, and indicate that support by listing it as their first choice cipher suite. If chacha20 boosting is enabled, we honor that preference for the sake of performance, rather than sticking to server order.

> The prioritization or 'boosting' of ChaCha20 ciphersuite preference is known as ChaCha20 boosting. If (and only if) an s2n-tls server selects a security policy with ChaCha20 boosting support AND the client has a ChaCha20 ciphersuite as its most preferred ciphersuite then does the server prioritize its ChaCha20 ciphersuites over its non-ChaCha20 ciphersuites.  

Reviewer 2: ```suggestion
The prioritization of ChaCha20 ciphersuites is known as ChaCha20 boosting. An s2n-tls server will prioritize its ChaCha20 ciphersuites over its non-ChaCha20 ciphersuites if the server configures a security policy with ChaCha20 boosting enabled AND the client sends a ChaCha20 ciphersuite as its most preferred ciphersuite.  
```
I would write this paragraph like this.

s2n-tls does not expose an API to control the order of preference for each ciphersuite or protocol version. s2n-tls follows the following order:

*NOTE*: All ChaCha20-Poly1305 cipher suites will not be available if s2n-tls is not built with an Openssl 1.1.1 libcrypto. The underlying encrypt/decrypt functions are not available in older versions.

1. Always prefer the highest protocol version supported
2. Always use forward secrecy where possible. Prefer ECDHE over DHE.
3. Prefer encryption ciphers in the following order: AES128, AES256, ChaCha20, 3DES, RC4.
4. Prefer record authentication modes in the following order: GCM, Poly1305, SHA256, SHA1, MD5.

> #### ChaCha20-Poly1305 Boosting
> 
> For hardware with AES-NI, then AES-based ciphers perform faster than their ChaCha20-Poly1305 counterparts. However, for hardware without AES-acceleration, ChaCha20 faster. In these cases, clients would prefer to negotiate an ChaCha20 ciphersuite for performance reasons by listing it as their first choice. 
> 
> In order to support this, s2n-tls supports ChaCha20-Poly1305 prioritization or ChaCha20-Poly1305 boosting. An s2n-tls server will prioritize its ChaCha20 ciphersuites over its non-ChaCha20 ciphersuites if the server configures a security policy with ChaCha20 boosting enabled AND the client sends a ChaCha20 ciphersuite as its most preferred ciphersuite. 

Reviewer: We shouldn't assume customers know what "AES-NI" is, and "AES-based ciphers" isn't a great phrase. We need to cover why ChaCha20 might be preferred, what the normal s2n-tls behavior would be, and what the boosted behavior is.

How about this?
```suggestion
#### ChaCha20 Boosting

s2n-tls usually prefers AES over ChaCha20. However, some clients-- particularly mobile or IOT devices-- do not support AES hardware acceleration, making AES less efficient and performant than ChaCha20. In this case, clients will indicate their preference for ChaCha20 by listing it first during cipher suite negotiation. Usually s2n-tls servers ignore client preferences, but s2n-tls offers "ChaCha20 boosted" security policies that will choose ChaCha20 over AES if the client indicates a preference for ChaCha20.
```

|   "20200207"   |      X       |      X     |   X    |
|   "rfc9151"    |              |      X     |        |

### ChaCha20-Poly1305 Boosting

> 

Author: Do you have a recommendation on where I can move this? If not, I can remove this section and put a small two-sentence explanation under the 'sec policy w/ chacha20 boosting' section.

> 

Reviewer 2: Adding a small explanation to the security policy section makes sense to me.
